### PR TITLE
hapi-auth-jwt2: update to hapi 17

### PR DESCRIPTION
**Pull request by SimonSchick**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [ ] The package does not provide its own types, and you can not add them.
- [ ] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [ ] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/dwyl/hapi-auth-jwt2
- [x] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

If removing a declaration:
- [ ] If a package was never on DefinitelyTyped, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

Pretty basic over all, did the callback to promise porting and added some missing stuff as well.
cc @jhsimms 

PLEASE IGNORE THE V7 folder!

### CODE REVIEW

**Comment by SimonSchick** [comment]
I noticed some severe documentation issues in the projects readme, I replaced some of the typings with the (correct) jsonwebtoken ones.

**Comment by SimonSchick** [comment]
See https://github.com/dwyl/hapi-auth-jwt2/issues/267

File: types/hapi-auth-jwt2/index.d.ts [hunk 1]
```diff
@@ -1,39 +1,43 @@
...
-     *
-     * @param err an internal error
-     * @param key the secret key
-     * @param extraInfo any additional information that you would like
-     * to use in `validateFunc` which can be accessed via
-     * `request.plugins['hapi-auth-jwt2'].extraInfo`
-     */
-    type KeyLookupCallback = (err: any, key: string, extraInfo?: any) => void;
+declare namespace hapiAuthJwt2 {
+    interface ExtraInfo {
```
Comment (armanio123) [inline types/hapi-auth-jwt2/index.d.ts]: Is generally a bad idea to have empty interface. Remove this.

File: types/hapi-auth-jwt2/index.d.ts [hunk 2]
```diff
@@ -42,41 +46,24 @@ declare namespace hapiAuthJwt2 {
         /**
          * The secret key used to check the signature of the token *or* a *key lookup function*
          */
-        key?: string | KeyLookup;
+        key?: string | string[] | Promise<{ isValid: boolean; key: string; extraInfo?: ExtraInfo }>;
```
Comment (armanio123) [inline types/hapi-auth-jwt2/index.d.ts]: Use "any" instead.

File: types/hapi-auth-jwt2/tslint.json [hunk 3]
```diff
@@ -1 +1,6 @@
-{ "extends": "dtslint/dt.json" }
+{
+    "extends": "dtslint/dt.json",
+    "rules": {
+        "no-empty-interface": false
```
Comment (armanio123) [inline types/hapi-auth-jwt2/tslint.json]: Don't disable lint rules.

**Review by armanio123 — CHANGES_REQUESTED** [review]

**Comment by armanio123** [comment]
@warrenseymour you're listed as an owner for this d.ts. Could you helps us reviewing it?

File: types/hapi-auth-jwt2/tslint.json (hunk 3 above)
Comment (SimonSchick) [reply types/hapi-auth-jwt2/tslint.json]: The reason for this is to to allow user enhancements.

File: types/hapi-auth-jwt2/index.d.ts (hunk 2 above)
Comment (SimonSchick) [reply types/hapi-auth-jwt2/index.d.ts]: No, that would reduce type safety.

File: types/hapi-auth-jwt2/index.d.ts (hunk 1 above)
Comment (SimonSchick) [reply types/hapi-auth-jwt2/index.d.ts]: See comment below.

File: types/hapi-auth-jwt2/tslint.json (hunk 3 above)
Comment (armanio123) [reply types/hapi-auth-jwt2/tslint.json]: That makes sense, ignore this comments.

**Review by armanio123 — APPROVED** [review]

**Pull request merged**